src/custom/pages/Swap/components/SwapButton/SwapButton.tsx

```diff
@@ -0,0 +1,169 @@
+import { SwapButtonState } from 'pages/Swap/helpers/useSwapButtonState'
+import { ReactNode } from 'react'
+import { ThemedText, ButtonSize } from 'theme'
+import { Trans } from '@lingui/macro'
+import { ButtonError, ButtonPrimary } from 'components/Button'
+import { Text } from 'rebass'
+import { Token } from '@uniswap/sdk-core'
+import { AutoRow } from 'components/Row'
+import { GreyCard } from 'components/Card'
+import { GpEther } from 'constants/tokens'
+import { SupportedChainId } from 'constants/chains'
+import { useToggleWalletModal } from 'state/application/hooks'
+import { AutoColumn } from 'components/Column'
+import { ApproveButton, ApproveButtonProps } from 'pages/Swap/components/ApproveButton'
+import * as styledEl from './styled'
+import { WrapType } from 'hooks/useWrapCallback'
+import { TransactionResponse } from '@ethersproject/providers'
+
+export interface SwapButtonProps {
+  swapButtonState: SwapButtonState
+  chainId: number | undefined
+  wrappedToken: Token
+  handleSwap: () => void
+  doSwap: () => void
+  approveButtonProps: ApproveButtonProps
+  onWrap?: () => Promise<TransactionResponse>
+  wrapType: WrapType
+  swapInputError?: ReactNode
+}
+
+export function SwapButton(props: SwapButtonProps) {
+  const {
+    swapButtonState,
+    chainId,
+    onWrap,
+    wrappedToken,
+    approveButtonProps,
+    swapInputError,
+    handleSwap,
+    doSwap,
+    wrapType,
+  } = props
+  const doWrap = () => onWrap && onWrap().catch((error) => console.error('Error ' + wrapType, error))
+  const toggleWalletModal = useToggleWalletModal()
+
+  const map: { [key in SwapButtonState]: JSX.Element } = {
```

Comment:
This map - and every entry in it - will be recreated on every re-render, right?
Wouldn't be better to use the same approach you used on the hook with `if`s for every state? Or a switch

Reply:
Thank you!
Fixed: https://github.com/cowprotocol/cowswap/pull/1049/commits/d6abd24f4765137aa91ee0fcb0e833cf194baf5e
I used the map here to ensure that all Swap button states are implemented. I can guarantee it using `[key in SwapButtonState]` type, if the map doesn't contain at least one of SwapButtonState then typescript will fail.

Reply:
But the same is true for a `switch`, isn't it?